plat/layerscape/common/aarch64/ls_console.S

```diff
@@ -0,0 +1,197 @@
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <arch.h>
+#include <asm_macros.S>
+#include <console_macros.S>
+#include <assert_macros.S>
+#include "ls_16550.h"
+	/*
+	 * "core" functions are low-level implementations that don't require
+	 * writable memory and are thus safe to call in BL1 crash context.
+	 */
+	.globl ls_cs_16550_core_init
+	.globl ls_cs_16550_core_putc
+	.globl ls_cs_16550_core_getc
+	.globl ls_cs_16550_core_flush
+
+	/* -----------------------------------------------
+	 * int ls_cs_16550_core_init(uintptr_t base_addr,
+	 * unsigned int uart_clk, unsigned int baud_rate)
+	 * Function to initialize the console without a
+	 * C Runtime to print debug information. This
+	 * function will be accessed by console_init and
+	 * crash reporting.
+	 * In: x0 - console base address
+	 *     w1 - Uart clock in Hz
+	 *     w2 - Baud rate
+	 * Out: return 1 on success, 0 on error
+	 * Clobber list : x1, x2, x3
+	 * -----------------------------------------------
+	 */
+func ls_cs_16550_core_init
+	/* Check the input base address */
+	cbz	x0, init_fail
+	/* Check baud rate and uart clock for sanity */
+	cbz	w1, init_fail
+	cbz	w2, init_fail
+
+	/* Program the baudrate */
+	/* Divisor =  Uart clock / (16 * baudrate) */
+	lsl	w2, w2, #4
+	udiv	w2, w1, w2
+	and	w1, w2, #0xff		/* w1 = DLL */
+	lsr	w2, w2, #8
+	and	w2, w2, #0xff		/* w2 = DLLM */
+	ldrb	w3, [x0, #UARTLCR]
+	orr	w3, w3, #UARTLCR_DLAB
+	strb	w3, [x0, #UARTLCR]	/* enable DLL, DLLM programming */
+	strb	w1, [x0, #UARTDLL]	/* program DLL */
+	strb	w2, [x0, #UARTDLLM]	/* program DLLM */
+	mov	w2, #~UARTLCR_DLAB
+	and	w3, w3, w2
+	strb	w3, [x0, #UARTLCR]	/* disable DLL, DLLM programming */
+
+	/* 8n1 */
+	mov	w3, #3
+	strb	w3, [x0, #UARTLCR]
+	/* no interrupt */
+	mov	w3, #0
+	strb	w3, [x0, #UARTIER]
+	/* enable fifo, DMA */
+	mov	w3, #(UARTFCR_FIFOEN |UARTFCR_TXCLR | UARTFCR_RXCLR)
+	strb	w3, [x0, #UARTFCR]
+	/* DTR + RTS */
+	mov	w3, #3
+	str	w3, [x0, #UARTMCR]
+	mov	w0, #1
+	ret
+init_fail:
+	mov	w0, #0
+	ret
+endfunc ls_cs_16550_core_init
+
+	.globl ls_cs_16550_register
+
+	/* -----------------------------------------------
+	 * int ls_cs_16550_register(console_16550_t *console,
+	 *	uintptr_t base, uint32_t clk, uint32_t baud)
+	 * Function to initialize and register a new 16550
+	 * console. Storage passed in for the console struct
+	 * *must* be persistent (i.e. not from the stack).
+	 * In: x0 - UART register base address
+	 *     w1 - UART clock in Hz
+	 *     w2 - Baud rate
+	 *     x3 - pointer to empty console_16550_t struct
+	 * Out: return 1 on success, 0 on error
+	 * Clobber list : x0, x1, x2, x6, x7, x14
+	 * -----------------------------------------------
+	 */
+func ls_cs_16550_register
+	mov	x7, x30
+	mov	x6, x3
+	cbz	x6, register_fail
+	str	x0, [x6, #CONSOLE_T_16550_BASE]
+
+	bl	ls_cs_16550_core_init
+	cbz	x0, register_fail
+
+	mov	x0, x6
+	mov	x30, x7
+	finish_console_register 16550
```

Comment:
With the current code, this macro is generating the names `console_16550_putc|getc|flush`, that don't exist. I don't know how this compiles, I can't see any reference to those names in `bl31.dump`. However, if you take a look at the dissasembly of `ls_cs_16550_register`, there are some weird loads:

```
00000000100058a8 <ls_cs_16550_register>:
    100058a8:	aa1e03e7 	mov	x7, x30
    100058ac:	aa0303e6 	mov	x6, x3
    100058b0:	b40001e6 	cbz	x6, 100058ec <register_fail>
    100058b4:	f90014c0 	str	x0, [x6,#40]
    100058b8:	97ffffe0 	bl	10005838 <ls_cs_16550_core_init>
    100058bc:	b4000180 	cbz	x0, 100058ec <register_fail>
    100058c0:	aa0603e0 	mov	x0, x6
    100058c4:	aa0703fe 	mov	x30, x7
    100058c8:	58000141 	ldr	x1, 100058f0 <register_fail+0x4>
    100058cc:	f9000801 	str	x1, [x0,#16]
    100058d0:	58000141 	ldr	x1, 100058f8 <register_fail+0xc>
    100058d4:	f9000c01 	str	x1, [x0,#24]
    100058d8:	58000141 	ldr	x1, 10005900 <register_fail+0x14>
    100058dc:	f9001001 	str	x1, [x0,#3
    100058e0:	d28000a1 	mov	x1, #0x5                   	// #5
    100058e4:	f9000401 	str	x1, [x0,#8]
    100058e8:	17ffff8f 	b	10005724 <console_register>
```

I think that the names you are supposed to use are `console_ls_16550_putc|getc|flush`. The `console` prefix is needed.

In this line, you will have to replace 16550 by `ls_16550`.